// @grafana/ui/dist/esm/components/InteractiveTable/Expander/index.mjs
import { jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { IconButton } from '../../IconButton/IconButton.mjs';

const expanderContainerStyles = css({
  display: "flex",
  alignItems: "center",
  height: "100%"
});
function ExpanderCell({ row, __rowID }) {
  return /* @__PURE__ */ jsx("div", { className: expanderContainerStyles, children: /* @__PURE__ */ jsx(
    IconButton,
    {
      tooltip: "toggle row expanded",
      "aria-controls": __rowID,
      name: row.isExpanded ? "angle-down" : "angle-right",
      "aria-expanded": row.isExpanded,
      ...row.getToggleRowExpandedProps(),
      size: "lg"
    }
  ) });
}
function ExpanderHeader({ isAllRowsExpanded, toggleAllRowsExpanded }) {
  return /* @__PURE__ */ jsx("div", { className: expanderContainerStyles, children: /* @__PURE__ */ jsx(
    IconButton,
    {
      "aria-label": !isAllRowsExpanded ? "Expand all rows" : "Collapse all rows",
      name: !isAllRowsExpanded ? "table-expand-all" : "table-collapse-all",
      onClick: () => toggleAllRowsExpanded(),
      size: "lg",
      tooltip: !isAllRowsExpanded ? "Expand all rows" : "Collapse all rows",
      variant: "secondary"
    }
  ) });
}

export { ExpanderCell, ExpanderHeader };
